Convert MaxTextureCacheCapacity to Dynamic MaxTextureCacheCapacity for High Resolution Mod support.

MaxLastBreath opened the pull request:
### Changes Context
- This change allows for games to boot and run at stable framerates at resolutions higher than native 4K.
- This also fixes crashes on games with modded native resolutions due to running out of texture cache.
- Potentially can improve Performance/Stuttering on games that previously were able to run at 4K or higher modded Resolutions.

### PR Changes:
- Added the Following variables to class AutoDeleteCache.
  - private const MaxTextureSizeCapacity (4 GiB).
  - private const MinTextureSizeCapacity (512 MiB) .
  - private const MemoryScaleFactor (0.50F).
  - private _maxCacheMemoryUsage.
- Added Initialize Function() in TextureCache and AutoDeleteCache.
  - Uses half of the current available Video Memory as TextureCache, clamped between Private Variables for Max and Min TextureSizeCapacity.
  - This function is called from MemoryManager constructor.
- Passed GPUContext into AutoDeleteCache.Initialize() from TextureCache.Initialize() class.
- Added a MaxGpuMemory variable in Capabilities structure.
- Added a Vulkan GPU Heap Iterator to get MaxGpuMemory (VulkanRenderer.cs GetTotalGPUMemory()).
- Set MaximumGpuMemory for OpenGL to 0.

### Edit:
- Comment edited to match all recent changes and add more context to PR.

GreemDev commented on src/Ryujinx.Graphics.Gpu/Image/AutoDeleteCache.cs:
```diff
@@ -55,12 +56,32 @@ class AutoDeleteCache : IEnumerable<Texture>
         /// Creates a new instance of the automatic deletion cache.
         /// </summary>
-        public AutoDeleteCache()
+        public AutoDeleteCache(GpuContext _Context)
```
C# styling dictates this should be `GpuContext context`; the underscore is present in the class field because it's a private readonly field.

MaxLastBreath replied:
Changed it 👍

MutantAura commented on src/Ryujinx.Graphics.Gpu/Image/AutoDeleteCache.cs:
```diff
@@ -46,7 +47,6 @@ class AutoDeleteCache : IEnumerable<Texture>
     {
         private const int MinCountForDeletion = 32;
         private const int MaxCapacity = 2048;
-        private const ulong MaxTextureSizeCapacity = 1024 * 1024 * 1024; // MB;
```
I think it would be best to keep this constant as `DefaultTextureSizeCapacity` to avoid the magic numbers present below.
Additionally I would add a constant for your chosen VRAM -> texture capacity scale factor (0.5f right now).

MaxLastBreath replied:
50% was the suggested ratio from GDK when i asked, this has to be further tested though but it seems pretty good so far from testing, for example i can comfortably run 6k and even 8K on my rtx 3080 on TOTK without ever running close to the vram cap :D. Question is how does it behave on lower end devices, like a steamdeck etc? Btw going to update stuff based on your suggestions now :)

MutantAura commented on src/Ryujinx.Graphics.Gpu/Image/AutoDeleteCache.cs:
```diff
@@ -56,11 +56,32 @@ class AutoDeleteCache : IEnumerable<Texture>
+        /// </summary>
+        private ulong GetMaxTextureCapacity()
+        {
+            Capabilities capabilities = _context.Capabilities;
+
+            if (capabilities.MaximumGpuMemory <= 0)
+            {
+                return 1024L * 1024 * 1024;
+            }
+            else
+            {
+                return (ulong)(capabilities.MaximumGpuMemory * 0.50);
+            }
```
I think this can be simplified to the following but its mainly aesthetic with some magic numbers removed.

```cs
ulong maxMemory = _context.Capabilities.MaximumGpuMemory;

if (maxMemory > 0)
{
    return (ulong)(maxMemory * MemoryScaleFactor)
}

return DefaultTextureSizeCapacity;
```

MutantAura commented on src/Ryujinx.Graphics.Vulkan/VulkanRenderer.cs:
```diff
@@ -781,7 +781,30 @@ public unsafe Capabilities GetCapabilities()
+
+        public ulong GetTotalGPUMemory()
+        {
+
+            ulong totalMemory = 0;
+
+
```
```cs
ulong totalMemory = 0
```

MutantAura replied:
Doesn't seem to have highlighted my intent here to delete the redundant newlines!

gdkchan commented on src/Ryujinx.Graphics.Vulkan/VulkanRenderer.cs:
```diff
@@ -781,7 +781,28 @@ public unsafe Capabilities GetCapabilities()
+                }
+            }
+
+            Logger.Info?.Print(LogClass.Gpu, $"GPU Memory: {totalMemory / (1024 * 1024)} MB");
```
If we're going to print this, it should be on `PrintGpuInformation` like the other GPU information, and it should likely use the `Notice` rather than `Info` log level.

gdkchan replied:
BTW, we currently use `MiB` rather than `MB`. (There was some discussion about changing that back to MB etc, but that's a separate issue).

gdkchan commented on src/Ryujinx.Graphics.Gpu/Image/AutoDeleteCache.cs:
```diff
@@ -56,11 +58,30 @@ class AutoDeleteCache : IEnumerable<Texture>
+        /// </summary>
+        private ulong GetMaxTextureCapacity()
+        {
+            ulong maxMemory = _context.Capabilities.MaximumGpuMemory;
```
Caching the calculated max memory locally would be slightly better still.
Like:
```cs
if (_maxCacheMemoryUsage == 0)
{
    // ...
    _maxCacheMemoryUsage = calculatedValue;
}

return _maxCacheMemoryUsage;
```

gdkchan replied:
An alternative is adding an `Initialize` method to the `TextureCache` and calling it from the `MemoryManager` constructor. At that point, `Capabilities` should have been set already. `AutoDeleteCache` would have an `Initialize` method too, called from the `TextureCache` and it would set `_maxCacheMemoryUsage`. `Add`, `Lift` etc would then be able to access `_maxCacheMemoryUsage` directly.

MaxLastBreath replied:
This is now done :)

MaxLastBreath commented:
### Ended up doing the following
- Added a new Const for MaxTextureSizeCapacity in AutoDeleteCache
- Added a new private var for _maxCacheMemoryUsage
- Added 2 new Initialize Functions just as suggested, one in TextureCache and one in AutoDeleteCache
- Added Notice level to the logger in VulkanRenderer and then moved it into PrintGpuInformation. As well as changed  MB to MiB
- Removed function GetMaxMemoryCache
- Opted to use a Ternary counter as opposed to if statements to determine the _maxCacheMemoryUsage

### Side Note:
- I think var maxMemory should probably be changed to something else, perhaps CalculatedCacheMemory? That seems a bit long.

MaxLastBreath commented on src/Ryujinx.Graphics.Gpu/Image/AutoDeleteCache.cs:
```diff
@@ -46,7 +47,8 @@ class AutoDeleteCache : IEnumerable<Texture>
         private const int MinCountForDeletion = 32;
         private const int MaxCapacity = 2048;
-        private const ulong MaxTextureSizeCapacity = 1024 * 1024 * 1024; // MB;
+        private const ulong DefaultTextureSizeCapacity = 1024 * 1024 * 1024;
```
Ended up putting a 4GB limit, however from what i heard from friends there may be benefits to a higher cache on some games like Metroid Prime Remastered, it seems to increase stability at high resolutions such as (5K+)? For now it seems rather easy enough to test with the new Values though. There's no actual reason to not have a cap, since eventually 8 GB DRAM would bottleneck before VRAM does, I think.

gdkchan commented on src/Ryujinx.Graphics.Gpu/Image/AutoDeleteCache.cs:
```diff
@@ -56,6 +60,19 @@ class AutoDeleteCache : IEnumerable<Texture>
+        {
+            var maxMemory = (ulong)(context.Capabilities.MaximumGpuMemory * MemoryScaleFactor);
+
+            _maxCacheMemoryUsage = maxMemory == 0 ?
```
Can you use a regular `if` here? IMO nested ternary ifs are not good for code readability.

gdkchan replied:
Alternatively, you could rewrite it as:
```cs
_maxCacheMemoryUsage = maxMemory == 0 ? DefaultTextureSizeCapacity : Math.Min(maxMemory, MaxTextureSizeCapacity);
```

gdkchan commented on src/Ryujinx.Graphics.Gpu/Image/AutoDeleteCache.cs:
```diff
@@ -56,6 +60,19 @@ class AutoDeleteCache : IEnumerable<Texture>
         private readonly Dictionary<TextureDescriptor, ShortTextureCacheEntry> _shortCacheLookup;
 
+        /// <summary>
+        /// Gets MaxTextureCapacity Dynamically
```
```suggestion
        /// Initializes the cache, setting the maximum texture capacity for the specified GPU context.
```
This needs to be updated since the function changed.
BTW, XML docs for the new `context` parameter is missing.

gdkchan commented on src/Ryujinx.Graphics.Gpu/Image/TextureCache.cs:
```diff
@@ -68,6 +68,11 @@ public TextureCache(GpuContext context, PhysicalMemory physicalMemory)
             _cache = new AutoDeleteCache();
         }
 
+        public void Initialize()
```
This needs XML docs too, it is fine to use more or less the same summary as the one on `AutoDeleteCache`.

gdkchan commented on src/Ryujinx.Graphics.Gpu/Image/AutoDeleteCache.cs:
```diff
@@ -56,6 +60,16 @@ class AutoDeleteCache : IEnumerable<Texture>
+        /// <summary>
+        /// Initializes the cache, setting the maximum texture capacity for the specified GPU context.
+        /// </summary>
+        public void Initialize(GpuContext context)
```
```suggestion
        <param name="context">The GPU context that the cache belongs to</param>
        public void Initialize(GpuContext context)
```

gdkchan commented on src/Ryujinx.Graphics.Vulkan/VulkanRenderer.cs:
```diff
@@ -781,7 +781,26 @@ public unsafe Capabilities GetCapabilities()
+                maximumGpuMemory: GetTotalGPUMemory());
+        }
+
+        public ulong GetTotalGPUMemory()
```
```suggestion
        private ulong GetTotalGPUMemory()
```

gdkchan reviewed (approved):
lgtm minus comment, thanks.

MaxLastBreath commented:
Public Build 4k 8192 Shadows
![Screenshot 2024-09-20 205158](https://github.com/user-attachments/assets/baac8279-7ddf-4144-9067-f13c0a9bed65)
PR Build 4k 8192 Shadows
![image](https://github.com/user-attachments/assets/b308f94d-03bb-4a6a-888e-a3a12ddc8e47)
Public Build 8k 8192 Shadows (1FPS titlescreen unplayable)
![image](https://github.com/user-attachments/assets/0038e0b4-3c34-404a-9499-ad20105dacbd)
PR Build 8k 8192 Shadows RTX 3080, pretty much playable 30 FPS.
![image](https://github.com/user-attachments/assets/ceaca7e8-751f-43f0-bac6-67600407c592)

There's improvements in other games too, but since I don't own many games I can't post screenshots.

gdkchan commented:
I guess this will now use a 512MB cache for OpenGL, which means that some 4K mods will no longer work on OpenGL right?

MaxLastBreath commented:
> I guess this will now use a 512MB cache for OpenGL, which means that some 4K mods will no longer work on OpenGL right?

I'm not sure, but that can easily be adjusted thou

riperiperi commented on src/Ryujinx.Graphics.Vulkan/VulkanRenderer.cs:
```diff
@@ -781,7 +781,26 @@ public unsafe Capabilities GetCapabilities()
+            for (int i = 0; i < memoryProperties.MemoryHeapCount; i++)
+            {
+                var heap = memoryProperties.MemoryHeaps[i];
+                if ((heap.Flags & MemoryHeapFlags.DeviceLocalBit) == MemoryHeapFlags.DeviceLocalBit)
```
I think it is possible for no heaps to have the device local flag, though I don't think the popular drivers are doing that right now.

MaxLastBreath replied:
Thank you, it was very informative, we could look into implementing a way to get the OpenGL available vram memory, but I'm unsure how to do that rn, otherwise the other option is to manually increase the memory back to the previous 1024?

Perharps using the TargetApi.OpenGL enum to set Memory to 1024, so there isn't an regression would be an okay idea for the moment?

riperiperi replied:
That should be fine, I think. I doubt any compatible GPU will have less than that.

riperiperi reviewed (approved):
I guess this is fine, but situations where this issue occurs should all involve the extended RAM option, so it's better tied to that. If a high res mod runs into this cache issue and _doesn't_ use extended RAM, then it is very likely using memory incorrectly and will cause a huge number of other issues if the texture overlaps other GPU memory or is partially mapped.

It could lead to large memory usage _at 1x_ if textures don't leave the cache some other way. It's really hard to reason about this as it requires manual testing with a bunch of logs to verify, but SMO when transitioning between kingdoms could have RAM usage patterns that _crash_ 8GB cards in the past.

I also don't like that the cache size has been changed for OpenGL.

gdkchan commented on src/Ryujinx.Graphics.Gpu/Image/AutoDeleteCache.cs:
```diff
@@ -56,6 +62,25 @@ class AutoDeleteCache : IEnumerable<Texture>
+
+            _maxCacheMemoryUsage = Math.Clamp(CacheMemory, MinTextureSizeCapacity, MaxTextureSizeCapacity);
+
+            if (context.Capabilities.Api == TargetApi.OpenGL)
```
Can't this just check if `context.Capabilities.MaximumGpuMemory` is zero? Then any backend that does not support getting the amount of VRAM can just set it to zero.

MaxLastBreath replied:
Roger that :)

OllieReynolds commented:
I have been using this dirty-build PR since the last few days. I no longer get random crashes when playing TOTK, so I am really happy with the results here. Gameplay seems much more stable. I am using Vulkan for renderer.

gdkchan commented on src/Ryujinx.Graphics.Gpu/Image/AutoDeleteCache.cs:
```diff
@@ -56,6 +61,25 @@ class AutoDeleteCache : IEnumerable<Texture>
+        /// <param name="context">The GPU context that the cache belongs to</param>
+        public void Initialize(GpuContext context)
+        {
+            var CacheMemory = (ulong)(context.Capabilities.MaximumGpuMemory * MemoryScaleFactor);
```
```suggestion
            var cacheMemory = (ulong)(context.Capabilities.MaximumGpuMemory * MemoryScaleFactor);
```

gdkchan commented on src/Ryujinx.Graphics.Gpu/Image/AutoDeleteCache.cs:
```diff
@@ -56,6 +61,25 @@ class AutoDeleteCache : IEnumerable<Texture>
+
+            _maxCacheMemoryUsage = Math.Clamp(CacheMemory, MinTextureSizeCapacity, MaxTextureSizeCapacity);
+
+            if (CacheMemory == 0)
```
```suggestion
            if (context.Capabilities.MaximumGpuMemory == 0)
```
I'd use that directly here. It's a bit weird to use the divided value.

gdkchan commented on src/Ryujinx.Graphics.Gpu/Image/AutoDeleteCache.cs:
```diff
@@ -56,6 +61,25 @@ class AutoDeleteCache : IEnumerable<Texture>
+        {
+            var CacheMemory = (ulong)(context.Capabilities.MaximumGpuMemory * MemoryScaleFactor);
+
+            _maxCacheMemoryUsage = Math.Clamp(CacheMemory, MinTextureSizeCapacity, MaxTextureSizeCapacity);
```
```suggestion
            _maxCacheMemoryUsage = Math.Clamp(cacheMemory, MinTextureSizeCapacity, MaxTextureSizeCapacity);
```
